data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,1032 @@
+            self.drt_medium.priority.to_string(),
+            self.drt_high.time_to_block.to_string(),
+            self.drt_high.priority.to_string(),
+            self.drt_opulent.time_to_block.to_string(),
+            self.drt_opulent.priority.to_string(),
+            self.vtt_stinky.time_to_block.to_string(),
+            self.vtt_stinky.priority.to_string(),
+            self.vtt_low.time_to_block.to_string(),
+            self.vtt_low.priority.to_string(),
+            self.vtt_medium.time_to_block.to_string(),
+            self.vtt_medium.priority.to_string(),
+            self.vtt_high.time_to_block.to_string(),
+            self.vtt_high.priority.to_string(),
+            self.vtt_opulent.time_to_block.to_string(),
+            self.vtt_opulent.priority.to_string(),
+        )
+    }
+}
+
+/// A estimate for priority and time-to-block.
+///
+/// Time-to-block states what is the expected time (in epochs) that it would take for a transaction
+/// with this priority to be included into a block.
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct PriorityEstimate {
+    pub priority: Priority,
+    pub time_to_block: TimeToBlock,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct TimeToBlock(u64);
+
+impl TimeToBlock {
+    pub fn from_secs(secs: u64) -> Self {
+        Self(secs)
+    }
+}
+
+impl fmt::Display for TimeToBlock {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&seconds_to_human_string(self.0))
+    }
+}
+
+#[inline]
+fn option_update(option: &mut Option<Priority>, candidate: Option<Priority>) {
+    match (&candidate, &option) {
+        (Some(new), Some(old)) if new < old => {
+            *option = candidate;
+        }
+        (Some(_), None) => {
+            *option = candidate;
+        }
+        _ => {}
+    }
+}
+
+/// Priority estimation strategies. To be used with `PriorityEngine::estimate_priority`.
+pub mod strategies {
+    use pdatastructs::topk::lossycounter::LossyCounter;
+
+    use super::*;
+
+    /// A priority estimation strategy that receives a list of targetted time-to-blocks expressed in
+    /// minutes and derives the priority from there by using a [lossy count algorithm].
+    ///
+    /// [lossy count algorithm]: https://en.wikipedia.org/wiki/Lossy_Count_Algorithm
+    #[allow(clippy::cast_precision_loss)]
+    #[allow(clippy::cast_possible_truncation)]
+    #[allow(clippy::cast_sign_loss)]
+    pub fn target_minutes<'a>(
+        priorities: impl IntoIterator<Item = &'a Priorities>,
+        target_minutes: [u16; 5],
+        seconds_per_epoch: u16,
+    ) -> PrioritiesEstimate {
+        // Make the priorities argument an iterator (if it was not already) and measure its length.
+        let priorities = priorities.into_iter();
+        let priorities_count = priorities
+            .size_hint()
+            .1
+            .unwrap_or(DEFAULT_QUEUE_CAPACITY_EPOCHS) as f64;
```

Comment:
`size_hint` shouldn't be used to get the number of elements in an iterator, but there is a `len` method from the [ExactSizeIterator](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html) trait, which is guaranteed to return the correct number of items, and setting the bound to `ExactSizeIterator` will make it a compile time error to call this function with an iterator that does not have a known size.

However I tried that and it doesn't work because the iterator returned by the `CircularQueue` does not implement `ExactSizeIterator`. And it should, because it is possible to know the number of elements in advance, right? So that, combined with the lack of a `get` method makes me think that this library is not suitable for production. So we should either find another similar library, or just use the `VecDeque` from the standard library which is basically an unbounded circular queue.

Reply:
I guess we should be good with `VecDeque::with_capacity` for as long as we guard the calls to `.push_front` with a check that will do `.pop_back` first if we're hitting the maximum capacity :thinking:

Reply:
That worked like a charm!